### PR TITLE
Teach photo-index about PhotoGroup

### PR DESCRIPTION
**Pull request by aboodman**
Ignore first two commits, they are under review separately.

### CODE REVIEW

**Comment by aboodman** [comment]
cc: @ehalpern 

This integrates `photo-index` with `photo-date-dedup`, a program that deduplicates using timestamps similarly to the way the UI used to work.

Once this is landed we'll want to add in the content-based deduping as well.

File: samples/go/photo-index/main.go [hunk 1]
```diff
@@ -197,10 +201,8 @@ func stringsByCount(db datas.Database, strings map[types.String]int) types.Map {
 
 func usage() {
 	fmt.Fprintf(os.Stderr, "photo-index indexes photos by common attributes\n\n")
-	fmt.Fprintf(os.Stderr, "Usage: %s -db=<db-spec> -out-ds=<name> [input-paths...]\n\n", path.Base(os.Args[0]))
-	fmt.Fprintf(os.Stderr, "  <db>             : Database to work with\n")
-	fmt.Fprintf(os.Stderr, "  <out-ds>         : Dataset to write index to\n")
-	fmt.Fprintf(os.Stderr, "  [input-paths...] : One or more paths within <db-spec> to crawl\n\n")
+	fmt.Fprintf(os.Stderr, "Usage: %s [flags] [input-paths...]\n\n", path.Base(os.Args[0]))
```
Comment (mikegray) [inline samples/go/photo-index/main.go]: Shouldn't out-ds still be here?

File: go/marshal/decode.go [hunk 2]
```diff
@@ -260,26 +261,37 @@ func nomsValueDecoder(v types.Value, rv reflect.Value) {
 	rv.Set(reflect.ValueOf(v))
 }
 
+func iterListOrSlice(v types.Value, t reflect.Type, f func(c types.Value, i uint64)) {
```
Comment (willhite) [inline go/marshal/decode.go]: Should the name of this be iterListOrSet()?

**Comment by zcstarr** [comment]
jenkins: test this

File: samples/go/photo-index/main.go [hunk 3]
```diff
@@ -28,8 +31,23 @@ func main() {
...
+
+type Date struct {
+	NsSinceEpoch float64
+}
+
 func index() (win bool) {
 	var dbStr = flag.String("db", "", "input database spec")
+	var groupsStr = flag.String("groups", "", "path within db to look for PhotoGroup structs")
 	var outDSStr = flag.String("out-ds", "", "output dataset to write to - if empty, defaults to input dataset")
 	verbose.RegisterVerboseFlags(flag.CommandLine)
```
Comment (willhite) [inline samples/go/photo-index/main.go]: Should this flag.Parse(false) be flag.Parse(true)?

**Comment by zcstarr** [comment]
jenkins: test this

File: samples/go/photo-index/main.go (hunk 1 above)
Comment (mikegray) [inline samples/go/photo-index/main.go]: ?????

**Pull request merged**